### ISSUE TITLE
Weave proxy unexpectedly closes the connection after /start request

### ISSUE THREAD

**Issue by 2opremio** [weaveworks/weave#1514]
Very similar to #1103 but not a duplicate.  In an ECS-optimized AMI in Amazon's EC2, `amazon-ecs-init start` fails when using the weave proxy (DOCKER_HOST=unix:///var/run/weave.sock)

```
2015-10-07T17:03:30Z [INFO] pre-start
2015-10-07T17:03:30Z [INFO] start
2015-10-07T17:03:30Z [INFO] Container name: /weaveproxy
2015-10-07T17:03:30Z [INFO] Container name: /weave
2015-10-07T17:03:30Z [INFO] Container name: /ecs-weave-ecs-demo-task-438-httpserver-defec08acae0b2a32e00
2015-10-07T17:03:30Z [INFO] Container name: /ecs-weave-ecs-demo-task-438-dataproducer-90f6e29c9a8fad9b1f00
2015-10-07T17:03:30Z [INFO] Container name: /weavescope
2015-10-07T17:03:30Z [INFO] No existing agent container to remove.
2015-10-07T17:03:30Z [INFO] Starting Amazon EC2 Container Service Agent
2015-10-07T17:03:31Z [ERROR] could not start Agent: persistent connection closed
```

This seems to have been introduced in Weave 1.1.1 (I am not able to reproduce it with Weave 1.1).

Here are the weaveproxy logs: https://gist.github.com/2opremio/59f5317189c26fb5fe67

Note the `Error forwarding chunked response body: write unix @: broken pipe`

Here is the communication between the `amazon-ecs-init`  (and `ecs-agent`) captured by `socat`: https://gist.github.com/2opremio/bc42d5d824e23d259c16

For future reference, here's how I sniffed the communication between `ecs-init` and the proxy:
```
sudo yum install -y socat
sudo mv /var/run/weave.sock /var/run/weave.sock.orig
sudo socat -t100  -v UNIX-LISTEN:/var/run/weave.sock,mode=777,reuseaddr,fork UNIX-CONNECT:/var/run/weave.sock.orig 2> /tmp/socat.out &
sudo start ecs
```

And finally, here's the same communication, but using the docker daemon directly: https://gist.github.com/2opremio/6c7f066865fb9b0e38b7

Here are some details about the docker version used:
```
$ docker version
Client version: 1.7.1
Client API version: 1.19
Go version (client): go1.4.2
Git commit (client): 786b29d/1.7.1
OS/Arch (client): linux/amd64
Server version: 1.7.1
Server API version: 1.19
Go version (server): go1.4.2
Git commit (server): 786b29d/1.7.1
OS/Arch (server): linux/amd64
```

Note how the Weave Proxy closes the connection after the `/start` request: https://gist.github.com/2opremio/bc42d5d824e23d259c16#file-socat-out-L123

whereas Docker doesn't: https://gist.github.com/2opremio/6c7f066865fb9b0e38b7#file-socat-docker-out-L119

**Comment by rade** [comment]
The only change to the proxy in 1.1.1 was #1476, which doesn't even change the code; only where the weave.sock lives. Are you sure that the only difference in your setup is the weave version 1.1.0 vs 1.1.1?

**Comment by 2opremio** [comment]
I'll double-check.

**Comment by rade** [comment]
The other difference between the 1.1.0 and 1.1.1 proxy is library versions, in particular docker-py.

**Comment by 2opremio** [comment]
The problem is not systematic but occurs very frequently with Weave 1.1.1 (18 out of 20 times).

I have tried again 20 times with 1.1.0 and I haven't been able to reproduce it once. Here's the socat session: https://gist.github.com/2opremio/06fbdd65b997dfcfddcd

**Comment by paulbellamy** [comment]
In this instance, is your container still being started correctly? Anything in the docker logs?

**Comment by 2opremio** [comment]
@paulbellamy Yep, the container (`ecs-agent`) is started, but the client (part of `eccs-init` which supervises `ecs-agent`) fails causing the supervisor to die (yeah, I know, see https://github.com/aws/amazon-ecs-init/issues/29 ).

I will reproduce again and get the docker daemon logs, but nothing suggests it's unhappy or doing anything wrong.

**Comment by 2opremio** [comment]
Yep, the docker daemon is perfectly happy:

```
time="2015-10-12T14:05:19.181193387Z" level=info msg="POST /v1.15/containers/create?hostconfig=%7B%22Binds%22%3A%5B%22%2Fvar%2Frun%2Fweave.sock%3A%2Fvar%2Frun%2Fdocker.sock%22%2C%22%2Fvar%2Flog%2Fecs%3A%2Flog%22%2C%22%2Fvar%2Flib%2Fecs%2Fdata%3A%2Fdata%22%2C%22%2Fetc%2Fecs%3A%2Fetc%2Fecs%22%2C
%22%2Fvar%2Fcache%2Fecs%3A%2Fvar%2Fcache%2Fecs%22%2C%22%2Fcgroup%3A%2Fcgroup%3Aro%22%2C%22%2Fvar%2Frun%2Fdocker%2Fexecdriver%3A%2Fvar%2Flib%2Fdocker%2Fexecdriver%3Aro%22%5D%2C%22PortBindings%22%3A%7B%2251678%2Ftcp%22%3A%5B%7B%22HostIP%22%3A%22127.0.0.1%22%2C%22HostPort%22%3A%2251678%22%7D%5D%7
D%2C%22RestartPolicy%22%3A%7B%7D%7D&name=ecs-agent" 
time="2015-10-12T14:05:19.439710702Z" level=info msg="POST /v1.15/containers/671c175b7a396bea68bb26714d2c9f9be3fc8e6ffeeac7142b11c0540d18f773/start" 
time="2015-10-12T14:05:19.555398680Z" level=info msg="GET /containers/671c175b7a396bea68bb26714d2c9f9be3fc8e6ffeeac7142b11c0540d18f773/json" 
```

**Comment by paulbellamy** [comment]
I've had a trawl through `net/http`, (as I believe we changed from go 1.4 to 1.5 between versions), but I don't see anything that is obviously racy... Based on the logs it looks like the proxy could be getting it's unix socket to docker dropped while trying to write the response, but that doesn't explain the error showing up twice. It could also potentially be an issue with the ContainerStarted event firing slowly, and teh docker response writer timing out.

**Comment by rade** [comment]
weave 1.1.1 was built with go 1.4.3.

**Comment by 2opremio** [comment]
@paulbellamy I have observed that the `write unix @: broken pipe` is not printed by the proxy every time I reproduce it. And, I am not completely sure, but it's possible that we observed the same behaviour in #1103 

My guess is that the client closes the connection right after receiving the unexpected `close`, sometimes causing a pending write to fail on the poxy's end.

In any case, I think that the main *problem* is the close.

**Comment by paulbellamy** [comment]
I believe I've traced it down to something in the build environment for 1.1.1, when I pull the image from dockerhub, I can reproduce it, but when I build tag v1.1.1 from source, I can't.

**Comment by paulbellamy** [comment]
It is reproducible with:
```sh
docker rm -f ecs-agent || true
curl -v -X POST --header "Content-Type: application/json" http://localhost:12375/v1.15/containers/create?name=ecs-agent --data-binary '{"Image":"alpine","Cmd":["/bin/sh    "]}'
curl -v -X POST --header "Content-Type: application/json" http://localhost:12375/v1.15/containers/ecs-agent/start --data-binary "null"
```
The last response will contain `Connection: close`

**Comment by rade** [comment]
Fails for me with a rebuilt v1.1.1.

Here are the versions of all my github packages:
```
$ for dir in $(go list -f '{{join .Deps "\n"}}' ./prog/weaveproxy | grep github | cut -d / -f 1-3 | sort | uniq) ; do pushd ~/go/src/$dir >/dev/null; echo $dir $(git rev-parse HEAD); popd >/dev/null; done
github.com/andybalholm/go-bit 34813fe0e46a87398d88d0315e0f22615592b242
github.com/docker/docker 072ac5e7f061aaece6693bb17510b3e968ac4971
github.com/fsouza/go-dockerclient af9789bbd78acf3e279274caa54682185eb7ed33
github.com/google/gopacket 63a65fadb1b52f38084c14d3eda50c57e2101cd1
github.com/gorilla/context 1c83b3eabd45b6d76072b66b746c20815fb2872d
github.com/gorilla/mux ad4d7a5882b961e07e2626045eb995c022ac6664
github.com/miekg/dns 874ec871288a738d8d87fd5cec1dd71e88fdacb2
github.com/opencontainers/runc 5d22c824be30afdf0146cdee44f6e9207cb1bfa9
github.com/Sirupsen/logrus 418b41d23a1bf978c06faea5313ba194650ac088
github.com/weaveworks/weave de95859e0d5b7b4a8d89e68e13c741c838d63797
$ go version
go version go1.4.3 linux/amd64
$ uname -a
Linux xps 3.16.0-50-generic #67~14.04.1-Ubuntu SMP Fri Oct 2 22:07:51 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by rade** [comment]
I did a `make update; git checkout v1.1.1; make clean; make` and repeated the test. Same result:

New dependency versions:
```
$ for dir in $(go list -f '{{join .Deps "\n"}}' ./prog/weaveproxy | grep github | cut -d / -f 1-3 | sort | uniq) ; do pushd ~/go/src/$dir >/dev/null; echo $dir $(git rev-parse HEAD); popd >/dev/null; done
github.com/andybalholm/go-bit 34813fe0e46a87398d88d0315e0f22615592b242
github.com/docker/docker 3856c5efa685013b5e82347d5812de4c7788d4ff
github.com/fsouza/go-dockerclient 09604abc82243886001c3f56fd709d4ba603cead
github.com/google/gopacket 1b0b78901cdd351ecfc68bf1a2adafcd2ff30220
github.com/gorilla/context 1c83b3eabd45b6d76072b66b746c20815fb2872d
github.com/gorilla/mux ad4d7a5882b961e07e2626045eb995c022ac6664
github.com/miekg/dns dfefa36553955889dbb9ef892d4bd116ec095ce2
github.com/opencontainers/runc d9ba9cebacf85b9361c7d1269c28d7fd686499e0
github.com/Sirupsen/logrus 4643a7efec9b8fd7548882e7b30095711343100f
github.com/weaveworks/weave de95859e0d5b7b4a8d89e68e13c741c838d63797
```

@paulbellamy I thought the test had worked, but the 'make update' had changed my weave checkout, so I built the wrong version. rebuilding v1.1.1 produced the same error as before.

**Comment by rade** [comment]
I've also just built v1.1.0 and got the same (failing) result. master works fine.

Another difference might be docker images, in particular gliderlabs/alpine and the packages we install on top of it in the weaveexec image. I completely wiped out all my docker images and did a fresh build of v1.1.0. The versions reported for the weavexec build are as follows:

```
sudo docker build -t weaveworks/weaveexec prog/weaveexec
Sending build context to Docker daemon 35.63 MB
Sending build context to Docker daemon 
Step 0 : FROM gliderlabs/alpine
latest: Pulling from gliderlabs/alpine
2cc966a5578a: Already exists 
Digest: sha256:8df853478e4601508b2f155313a26235ba65c96db293d8bc6e1034fa84bf7025
Status: Downloaded newer image for gliderlabs/alpine:latest
 ---> 2cc966a5578a
Step 1 : MAINTAINER Weaveworks Inc <help@weave.works>
 ---> Running in 0cb56e71bbf3
 ---> a04a6fd2b904
Removing intermediate container 0cb56e71bbf3
Step 2 : LABEL works.weave.role system
 ---> Running in d6d06f41f5bb
 ---> b56e817756c8
Removing intermediate container d6d06f41f5bb
Step 3 : WORKDIR /home/weave
 ---> Running in 583bbbbb3428
 ---> 335ecf6242bc
Removing intermediate container 583bbbbb3428
Step 4 : VOLUME /w
 ---> Running in 1dd31cc96587
 ---> 1055fc5062c8
Removing intermediate container 1dd31cc96587
Step 5 : ENTRYPOINT /home/weave/sigproxy /home/weave/weave
 ---> Running in 35408ebda746
 ---> a7f117066725
Removing intermediate container 35408ebda746
Step 6 : RUN apk add --update     curl     ethtool     iptables     iproute2     util-linux     conntrack-tools     bind-tools   && rm -rf /var/cache/apk/*
 ---> Running in 31b1c510e39c
fetch http://alpine.gliderlabs.com/alpine/v3.2/main/x86_64/APKINDEX.tar.gz
(1/35) Installing libattr (2.4.47-r3)
(2/35) Installing libcap (2.24-r1)
(3/35) Installing libgcc (4.9.2-r5)
(4/35) Installing bind-libs (9.10.2_p4-r0)
(5/35) Installing bind-tools (9.10.2_p4-r0)
(6/35) Installing libmnl (1.0.3-r0)
(7/35) Installing libnfnetlink (1.0.1-r1)
(8/35) Installing libnetfilter_conntrack (1.0.4-r0)
(9/35) Installing libnetfilter_cthelper (1.0.0-r0)
(10/35) Installing libnetfilter_cttimeout (1.0.0-r0)
(11/35) Installing libnetfilter_queue (1.0.2-r0)
(12/35) Installing libcom_err (1.42.13-r0)
(13/35) Installing krb5-conf (1.0-r0)
(14/35) Installing keyutils-libs (1.5.9-r1)
(15/35) Installing libverto (0.2.5-r0)
(16/35) Installing krb5-libs (1.13.1-r1)
(17/35) Installing libtirpc (0.3.0-r1)
(18/35) Installing conntrack-tools (1.4.3-r0)
(19/35) Installing run-parts (4.4-r0)
(20/35) Installing openssl (1.0.2d-r0)
(21/35) Installing lua5.2-libs (5.2.4-r0)
(22/35) Installing lua5.2 (5.2.4-r0)
(23/35) Installing ncurses-terminfo-base (5.9-r3)
(24/35) Installing ncurses-widec-libs (5.9-r3)
(25/35) Installing lua5.2-posix (33.3.1-r2)
(26/35) Installing ca-certificates (20141019-r2)
(27/35) Installing libssh2 (1.5.0-r0)
(28/35) Installing curl (7.42.1-r0)
(29/35) Installing ethtool (3.18-r0)
(30/35) Installing iptables (1.4.21-r3)
(31/35) Installing iproute2 (3.18.0-r0)
Executing iproute2-3.18.0-r0.post-install
(32/35) Installing libuuid (2.26.2-r0)
(33/35) Installing libblkid (2.26.2-r0)
(34/35) Installing libmount (2.26.2-r0)
(35/35) Installing util-linux (2.26.2-r0)
Executing busybox-1.23.2-r0.trigger
Executing ca-certificates-20141019-r2.trigger
OK: 21 MiB in 50 packages
 ---> c050f3df248e
Removing intermediate container 31b1c510e39c
Step 7 : ADD ./weave ./sigproxy ./weaveproxy /home/weave/
 ---> c0af26d7b867
Removing intermediate container f35682c1ec04
Step 8 : ADD ./netcheck ./weavehosts /usr/bin/
 ---> 423473118fdc
Removing intermediate container 68ec19e514f3
Step 9 : ADD ./weavewait /w/w
 ---> 5b37cf5a5975
Removing intermediate container 3afbd6f5496c
Step 10 : ADD ./docker.tgz /
 ---> 5c9f3c4d0171
Removing intermediate container 25eb0af15e4a
Successfully built 5c9f3c4d0171
```

**Comment by rade** [comment]
It's a go version issue!!!

I have just rebuilt with go 1.4.2 and the test works fine.

Looking at the diff between 1.4.2 and 1.4.3, there is certainly a [suspicious change](https://github.com/golang/go/commit/cb65428710d70abdaf101defa9cd7eaddff9d925).

Would be good if somebody else confirmed my findings.

**Comment by paulbellamy** [comment]
It was fixed here: https://github.com/golang/go/blob/master/src/net/http/server.go#L917

We should close this as not_our_problem, as it's fixed in Go 1.5.

**Comment by rade** [comment]
> https://github.com/golang/go/blob/master/src/net/http/server.go#L917

That is part of https://github.com/golang/go/commit/c2db5f4ccc61ba7df96a747e268a277b802cbb87, which is what was supposedly backported to 1.4 in the change that causes the problem. Looks like they just missed the `ErrBodyReadAfterClose` case in the backport!

>  it's fixed in Go 1.5 (maybe 1.5.1).

1.5

**Comment by bboreham** [comment]
Note release 1.1.2 is built with Go 1.4.2 in order to avoid this issue.
